src/main/java/sparksoniq/spark/ml/AnnotateFunctionIterator.java

```diff
@@ -0,0 +1,67 @@
+package sparksoniq.spark.ml;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.rumbledb.api.Item;
+import org.rumbledb.exceptions.ExceptionMetadata;
+import org.rumbledb.exceptions.UnexpectedTypeException;
+import org.rumbledb.items.ObjectItem;
+import org.rumbledb.runtime.DataFrameRuntimeIterator;
+import org.rumbledb.runtime.RuntimeIterator;
+import sparksoniq.jsoniq.ExecutionMode;
+import sparksoniq.semantics.DynamicContext;
+import sparksoniq.spark.DataFrameUtils;
+
+import java.util.List;
+
+public class AnnotateFunctionIterator extends DataFrameRuntimeIterator {
+
+    private static final long serialVersionUID = 1L;
+
+    public AnnotateFunctionIterator(
+            List<RuntimeIterator> arguments,
+            ExecutionMode executionMode,
+            ExceptionMetadata iteratorMetadata
+    ) {
+        super(arguments, executionMode, iteratorMetadata);
+    }
+
+    @Override
+    public Dataset<Row> getDataFrame(DynamicContext context) {
+        RuntimeIterator inputDataIterator = this.children.get(0);
+        RuntimeIterator schemaIterator = this.children.get(1);
+        ObjectItem schemaItem;
+
+        // materialize singleton pattern
```

Comment:
Do we not have a materialize function for this that works out of the box? Any reason why we cannot use it?

Reply:
As far as I know we have two materialize methods: one that materializes everything and another one that materializes the first result if it exists (otherwise returns null). The 'materialize singleton' as I call it has not been implemented yet. That is, materializes a singleton result and throws an error result set is not a singleton(either empty or multiple). Once implemented this can be used in multiple places across the project.

Regarding the signature and type-checks, I can take your word for them as I have not extensively worked with them. I have a feeling that since these mechanisms are in place, we can be safely remove manual checks that exist in the codebase in the near future. For the time being however, to be consistent with all the other iterators that implement these manual checks, I feel we should add these to the annotate function too. What do you think ?

Reply:
Yes, it's fine to keep them for now, we can still remove later.

For the materialize method, if it is not implemented, it suffices to open an issue.